src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -1555,32 +1539,66 @@ private bool ExactArrayInference(TypeSymbol source, TypeSymbol target, ref HashS
                 return false;
             }
 
-            var arraySource = (ArrayTypeSymbol)source;
-            var arrayTarget = (ArrayTypeSymbol)target;
+            var arraySource = (ArrayTypeSymbol)source.TypeSymbol;
+            var arrayTarget = (ArrayTypeSymbol)target.TypeSymbol;
             if (!arraySource.HasSameShapeAs(arrayTarget))
             {
                 return false;
             }
 
-            ExactInference(arraySource.ElementType.TypeSymbol, arrayTarget.ElementType.TypeSymbol, ref useSiteDiagnostics);
+            ExactInference(arraySource.ElementType, arrayTarget.ElementType, ref useSiteDiagnostics);
             return true;
         }
 
-        private bool ExactNullableInference(TypeSymbol source, TypeSymbol target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private enum ExactOrBoundsKind
+        {
+            Exact,
+            LowerBound,
+            UpperBound,
+        }
+
+        private void ExactOrBoundsInference(ExactOrBoundsKind kind, TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            switch (kind)
+            {
+                case ExactOrBoundsKind.Exact:
+                    ExactInference(source, target, ref useSiteDiagnostics);
+                    break;
+                case ExactOrBoundsKind.LowerBound:
+                    LowerBoundInference(source, target, ref useSiteDiagnostics);
+                    break;
+                case ExactOrBoundsKind.UpperBound:
+                    UpperBoundInference(source, target, ref useSiteDiagnostics);
+                    break;
+            }
+        }
+
+        private bool ExactOrBoundsNullableInference(ExactOrBoundsKind kind, TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations target, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
             Debug.Assert((object)source != null);
             Debug.Assert((object)target != null);
 
-            if (!source.IsNullableType() || !target.IsNullableType())
+            if (source.IsNullableType() && target.IsNullableType())
```

Comment:
Consider adding a comment to clarify what the two conditions handle (assuming they could not be unified).